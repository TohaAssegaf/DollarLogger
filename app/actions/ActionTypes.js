import {
    Payment
} from '/app/store/state/PaymentState'

// Set budget total.
export const SET_BUDGET_TOTAL_REQUEST = 'SET_BUDGET_TOTAL_REQUEST'
export type SetBudgetTotalRequestAction = {
    type: SET_BUDGET_TOTAL_REQUEST
}
export const SET_BUDGET_TOTAL_SUCCESS = 'SET_BUDGET_TOTAL_SUCCESS'
export type SetBudgetTotalSuccessAction = {
    type: SET_BUDGET_TOTAL_SUCCESS,
    total: number
}
export const SET_BUDGET_TOTAL_FAILURE = 'SET_BUDGET_TOTAL_FAILURE'
export type SetBudgetTotalFailureAction = {
    type: SET_BUDGET_TOTAL_FAILURE,
    errorMessage: string
}

// Get budget total.
export const GET_BUDGET_TOTAL_REQUEST = 'GET_BUDGET_TOTAL_REQUEST'
export type GetBudgetTotalRequestAction = {
    type: GET_BUDGET_TOTAL_REQUEST
}
export const GET_BUDGET_TOTAL_SUCCESS = 'GET_BUDGET_TOTAL_SUCCESS'
export type GetBudgetTotalSuccessAction = {
    type: GET_BUDGET_TOTAL_SUCCESS,
    total: number
}
export const GET_BUDGET_TOTAL_FAILURE = 'GET_BUDGET_TOTAL_FAILURE'
export type GetBudgetTotalFailureAction = {
    type: GET_BUDGET_TOTAL_FAILURE,
    errorMessage: string
}

// Create payment.
export const CREATE_PAYMENT_REQUEST = 'CREATE_PAYMENT_REQUEST'
export type CreatePaymentRequestAction = {
    type: CREATE_PAYMENT_REQUEST
}
export const CREATE_PAYMENT_SUCCESS = 'CREATE_PAYMENT_SUCCESS'
export type CreatePaymentSuccessAction = {
    type: CREATE_PAYMENT_SUCCESS,
    payment: Payment
}
export const CREATE_PAYMENT_FAILURE = 'CREATE_PAYMENT_FAILURE'
export type CreatePaymentFailureAction = {
    type: CREATE_PAYMENT_FAILURE,
    errorMessage: string
}

export type Action = |
    SetBudgetTotalRequestAction |
    SetBudgetTotalSuccessAction |
    SetBudgetTotalFailureAction |
    GetBudgetTotalRequestAction |
    GetBudgetTotalSuccessAction |
    GetBudgetTotalFailureAction |
    CreatePaymentRequestAction |
    CreatePaymentSuccessAction |
    CreatePaymentFailureAction