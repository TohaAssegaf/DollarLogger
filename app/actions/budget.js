import {
  SET_BUDGET_TOTAL_REQUEST,
  SET_BUDGET_TOTAL_SUCCESS,
  SET_BUDGET_TOTAL_FAILURE,
  GET_BUDGET_TOTAL_REQUEST,
  GET_BUDGET_TOTAL_SUCCESS,
  GET_BUDGET_TOTAL_FAILURE,
  SetBudgetTotalRequestAction,
  SetBudgetTotalSuccessAction,
  SetBudgetTotalFailureAction,
  GetBudgetTotalRequestAction,
  GetBudgetTotalSuccessAction,
  GetBudgetTotalFailureAction
} from './ActionTypes'
import * as BudgetModel from '/app/store/models/budget'

export function setBudgetTotalRequest(): SetBudgetTotalRequestAction {
  return {
    type: SET_BUDGET_TOTAL_REQUEST
  }
}

export function setBudgetTotalSuccess(total: number): SetBudgetTotalSuccessAction {
  return {
    type: SET_BUDGET_TOTAL_SUCCESS,
    total
  }
}

export function setBudgetTotalFailure(errorMessage: string): SetBudgetTotalFailureAction {
  return {
    type: SET_BUDGET_TOTAL_FAILURE,
    errorMessage
  }
}

export function setBudgetTotal(total: number) {
  return function(dispatch) {
    if (total <= 0) {
      dispatch(setBudgetTotalFailure("Budget must be greater than 0"))
    }
    if (!Number.isInteger(total)) {
      dispatch(setBudgetTotalFailure("Budget must be an integer"))
    }
    dispatch(setBudgetTotalRequest(total))
    return BudgetModel.setTotal(total)
      .then(() => {
        dispatch(setBudgetTotalSuccess(total))
        dispatch(NavigationActions.back())
      })
      .catch(error => dispatch(setBudgetTotalFailure(error.message)))
  }
}

export function getBudgetTotalRequest(): GetBudgetTotalRequestAction {
  return {
    type: GET_BUDGET_TOTAL_REQUEST
  }
}

export function getBudgetTotalSuccess(total: ?number): GetBudgetTotalSuccessAction {
  return {
    type: GET_BUDGET_TOTAL_SUCCESS,
    total
  }
}

export function getBudgetTotalFailure(errorMessage: string): GetBudgetTotalFailureAction {
  return {
    type: GET_BUDGET_TOTAL_FAILURE,
    errorMessage
  }
}

export function getBudgetTotal() {
  return function(dispatch) {
    dispatch(getBudgetTotalRequest())
    return BudgetModel.getTotal()
      .then(total => {
        dispatch(getBudgetTotalSuccess(total))
      })
      .catch(error => dispatch(getBudgetTotalFailure(error.message)))
  }
}
