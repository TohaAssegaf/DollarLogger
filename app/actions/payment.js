import {
  CREATE_PAYMENT_REQUEST,
  CREATE_PAYMENT_SUCCESS,
  CREATE_PAYMENT_FAILURE,
  CreatePaymentRequestAction,
  CreatePaymentSuccessAction,
  CreatePaymentFailureAction
} from './ActionTypes'
import * as PaymentModel from '/app/store/models/payment'
import {
  Payment
} from '/app/store/state/PaymentState'

export function createPaymentRequest(): CreatePaymentRequestAction {
  return {
    type: CREATE_PAYMENT_REQUEST
  }
}

export function createPaymentSuccess(payment: Payment): CreatePaymentSuccessAction {
  return {
    type: CREATE_PAYMENT_SUCCESS,
    payment
  }
}

export function createPaymentFailure(errorMessage: string): CreatePaymentFailureAction {
  return {
    type: CREATE_PAYMENT_FAILURE,
    errorMessage
  }
}

export function createPayment(total: number, name: string, date: Date) {
  return function (dispatch) {
    if (!Number.isInteger(total)) {
      dispatch(createPaymentFailure("Payment total must be an integer"))
    }
    dispatch(createPaymentRequest())
    return PaymentModel.createPayment(total, name, date)
      .then(payment => dispatch(createPaymentSuccess(payment)))
      .catch(error => dispatch(createPaymentFailure(error.message)))
  }
}