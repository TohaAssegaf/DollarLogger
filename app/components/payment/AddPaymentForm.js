import BasePaymentForm from './BasePaymentForm'
import actions from '~/app/actions'
import PaymentBuilder from '~/app/lib/PaymentBuilder'
import React from 'react'
import { connect } from 'react-redux'

class AddPaymentForm extends React.Component {
  createPayment(total, name, date, splitCount) {
    this.props.createPayment(
      new PaymentBuilder()
        .setName(name)
        .setTotal(total)
        .setDate(date)
        .setSplitCount(splitCount)
        .build())

    // Temporary hack. This should check state with a componentDidUpdate for write to be complete.
    this.props.navigation.goBack()
  }

  render() {
    return (
      <BasePaymentForm
        onSubmit={
          (total, name, date, splitCount) => this.createPayment(total, name, date, splitCount)}
      />
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    createPayment: (payment) => {
      dispatch(actions.createPayment(payment))
    }
  }
}

export default connect((state) => ({}), mapDispatchToProps)(AddPaymentForm)
