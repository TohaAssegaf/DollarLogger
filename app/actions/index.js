import * as budget from './budget'

export default Object.assign({}, budget)
