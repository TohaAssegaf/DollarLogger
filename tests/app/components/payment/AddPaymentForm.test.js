import actions from '~/app/actions'
import AddPaymentForm from '~/app/components/payment/AddPaymentForm'
import enzyme from 'enzyme'
import { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import MockDate from 'mockdate'
import React from 'react'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'

enzyme.configure({ adapter: new Adapter() })
const navigation = { goBack: jest.fn() }
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

it('renders correctly', () => {
  MockDate.set(new Date(Date.UTC(2018, 4, 2)))
  const store = mockStore({})
  const rendered = renderer.create(<AddPaymentForm store={store} navigation={navigation} />)
      .toJSON();
  expect(rendered).toMatchSnapshot();
});

it('dispatches add payment action and navigates back', () => {
  const store = mockStore({})
  const wrapper = shallow(<AddPaymentForm store={store} navigation={navigation} />)
      .dive({ context: { store } })
  const expectedTotal = 10000
  const expectedName = "Test payment"
  const expectedDate = new Date(2018, 4, 2)

  wrapper.simulate('submit', expectedTotal, expectedName, expectedDate)

  expect(store.getActions()).toContainEqual(actions.getPaymentsRequest())
  expect(navigation.goBack.mock.calls).toHaveLength(1)
});
