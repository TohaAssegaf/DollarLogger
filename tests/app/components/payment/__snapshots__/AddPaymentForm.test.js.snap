// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 20,
            },
            Object {
              "fontSize": 25,
            },
          ]
        }
      >
        $
      </Text>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onBlur={[Function]}
        onChangeText={[Function]}
        style={
          Array [
            Object {
              "fontSize": 20,
              "minWidth": 100,
            },
            Object {
              "fontSize": 25,
            },
          ]
        }
        value={undefined}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChange={[Function]}
      placeholder="Name"
      style={
        Object {
          "fontSize": 25,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "width": 142,
          },
          Object {
            "width": "auto",
          },
        ]
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
            },
            Object {
              "alignItems": "flex-start",
              "height": "auto",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "#aaa",
                "borderWidth": 1,
                "flex": 1,
                "height": 40,
                "justifyContent": "center",
              },
              Object {
                "alignItems": "flex-start",
                "borderWidth": 0,
                "height": "auto",
              },
              false,
              false,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#333",
                },
                Object {
                  "fontSize": 25,
                },
              ]
            }
          >
            05/02/2018
          </Text>
        </View>
        <Modal
          animationType="none"
          hardwareAccelerated={false}
          onRequestClose={[Function]}
          supportedOrientations={
            Array [
              "portrait",
              "portrait-upside-down",
              "landscape",
              "landscape-left",
              "landscape-right",
            ]
          }
          transparent={true}
          visible={false}
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "backgroundColor": "#00000077",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                accessibilityComponentType={undefined}
                accessibilityLabel={undefined}
                accessibilityTraits={undefined}
                accessible={true}
                hasTVPreferredFocus={undefined}
                hitSlop={undefined}
                isTVSelectable={true}
                nativeID={undefined}
                onLayout={undefined}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "flex": 1,
                  }
                }
                testID={undefined}
                tvParallaxProperties={undefined}
              >
                <View
                  collapsable={undefined}
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "height": 0,
                      "overflow": "hidden",
                    }
                  }
                >
                  <View
                    pointerEvents="auto"
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderTopColor": "#ccc",
                            "borderTopWidth": 1,
                            "marginTop": 42,
                          },
                          undefined,
                        ]
                      }
                    >
                      <RCTDatePicker
                        date={1525244400000}
                        locale={undefined}
                        maximumDate={undefined}
                        minimumDate={undefined}
                        minuteInterval={undefined}
                        mode="date"
                        onChange={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Object {
                            "height": 216,
                          }
                        }
                        timeZoneOffsetInMinutes={undefined}
                      />
                    </View>
                  </View>
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "height": 42,
                          "justifyContent": "center",
                          "padding": 20,
                          "position": "absolute",
                          "top": 0,
                        },
                        Object {
                          "left": 0,
                        },
                        undefined,
                      ]
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <Text
                      accessible={true}
                      allowFontScaling={true}
                      ellipsizeMode="tail"
                      style={
                        Array [
                          Object {
                            "color": "#46cf98",
                            "fontSize": 16,
                          },
                          Object {
                            "color": "#666",
                          },
                          undefined,
                        ]
                      }
                    >
                      取消
                    </Text>
                  </View>
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "height": 42,
                          "justifyContent": "center",
                          "padding": 20,
                          "position": "absolute",
                          "top": 0,
                        },
                        Object {
                          "right": 0,
                        },
                        undefined,
                      ]
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <Text
                      accessible={true}
                      allowFontScaling={true}
                      ellipsizeMode="tail"
                      style={
                        Array [
                          Object {
                            "color": "#46cf98",
                            "fontSize": 16,
                          },
                          undefined,
                        ]
                      }
                    >
                      确定
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </Modal>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "marginHorizontal": 20,
        "marginTop": 20,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel="Save new payment"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#6A1B9A",
              },
            ]
          }
        >
          SAVE
        </Text>
      </View>
    </View>
  </View>
</View>
`;
